README.md

# node-diablo2-data

<!-- PROJECT SHIELDS -->
[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]
[![NPM version][npm-shield]][npm-url]
[![Build Status][build-shield]][build-url]
[![Discord][discord-shield]][discord-url]

<!-- TABLE OF CONTENTS -->
## Table of Contents

* [About the Project](#about-the-project)
* [Getting Started](#getting-started)
  * [Prerequisites](#prerequisites)
  * [Installation](#installation)
* [Usage](#usage)
* [Roadmap](#roadmap)
* [Contributing](#contributing)
* [License](#license)
* [Acknowledgements](#acknowledgements)

<!-- ABOUT THE PROJECT -->
## About The Project

The goal of this project is to provide an easy-to-use lib to get Diablo 2 related data

## Getting Started

This is an example of how you may give instructions on setting up your project locally.
To get a local copy up and running follow these simple example steps.

### Prerequisites

* nodejs, npm

### Installation

```bash
npm i diablo2-data
```

<!-- USAGE EXAMPLES -->
## Usage

```js
const diabloData = require('diablo2-data')('pod_1.13d')

console.log(JSON.stringify(diabloData.skills[50]))
console.log(JSON.stringify(diabloData.areas[101]))
console.log(JSON.stringify(diabloData.npcsByName['Akara']))
console.log(JSON.stringify(diabloData.objectsByName['StoneTheta']))
console.log(JSON.stringify(diabloData.itemEnums.Action))
console.log(JSON.stringify(diabloData.itemPropertiesByName['Maximum Damage']))
console.log(JSON.stringify(diabloData.itemDataByName['Ring Mail']))
console.log(JSON.stringify(diabloData.itemTypesByName['Amulet']))
console.log(JSON.stringify(diabloData.uniqueItemsByName['Shako']))
console.log(JSON.stringify(diabloData.responses[41]))
```

Or run the example

```bash
sudo node examples/basic.js
```

## Roadmap

See the [open issues](https://github.com/Mephistools/node-diablo2-data/issues) for a list of proposed features (and known issues).

## Contributing

Contributions are what make the open source community such an amazing place to be learn, inspire, and create. Any contributions you make are **greatly appreciated**.

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request

## License

Distributed under the MIT License. See `LICENSE` for more information.

## Acknowledgements

* [Img Shields](https://shields.io)
* [Choose an Open Source License](https://choosealicense.com)
* [GitHub Pages](https://pages.github.com)

<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
[contributors-shield]: https://img.shields.io/github/contributors/Mephistools/node-diablo2-data.svg?style=flat-square
[contributors-url]: https://github.com/Mephistools/node-diablo2-data/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/Mephistools/node-diablo2-data.svg?style=flat-square
[forks-url]: https://github.com/Mephistools/node-diablo2-data/network/members
[stars-shield]: https://img.shields.io/github/stars/Mephistools/node-diablo2-data.svg?style=flat-square
[stars-url]: https://github.com/Mephistools/node-diablo2-data/stargazers
[issues-shield]: https://img.shields.io/github/issues/Mephistools/node-diablo2-data.svg?style=flat-square
[issues-url]: https://github.com/Mephistools/node-diablo2-data/issues
[license-shield]: https://img.shields.io/github/license/Mephistools/node-diablo2-data.svg?style=flat-square
[license-url]: https://github.com/Mephistools/node-diablo2-data/blob/master/LICENSE.txt
[npm-shield]: https://img.shields.io/npm/v/diablo2-data.svg
[npm-url]: http://npmjs.com/package/diablo2-data
[build-shield]: https://github.com/Mephistools/node-diablo2-data/workflows/CI/badge.svg
[build-url]: https://github.com/Mephistools/node-diablo2-data/actions?query=workflow%3A%22CI%22
[discord-shield]: https://img.shields.io/badge/chat-on%20discord-brightgreen.svg
[discord-url]: https://discord.gg/9RqtApv


docs/HISTORY.md
## 1.6.0
* Add server responses enums 

## 1.5.0
* Add stash bottom item enum container

## 1.4.0
* Implement inventory extension enum (pod only)

## 1.3.0
* Add item enums (move from diablo2-protocol)

## 1.2.0
* Add belts

## 1.1.0
* Add itemTypes
* Add qualityItems
* Add uniqueItems
* Add weapons
* Add superUniques
* Add setItems
* Add runes
* Add misc
* Add monStats
* Add missiles
* Add armor
* Add gems

## 1.0.0

* basic structure is there, some data is available for use in clients


index.js
const diabloDataToNode = require('./lib/loader')
const fs = require('fs')

function loadCsv (path, propertiesDelimiter, customHeaders) {
  const data = fs.readFileSync(path, 'utf8')
  const lines = data.split('\n')
  let headers
  if (customHeaders === undefined) {
    headers = lines.shift().split('\t')
  } else {
    headers = customHeaders
  }
  return lines
    .map(line => line.split(propertiesDelimiter))
    .map(arr => arr.reduce((acc, e, i) => {
      acc[headers[i]] = e
      return acc
    }, {}))
}

module.exports = function (diabloVersion) {
  const itemEnums = require(`./data/${diabloVersion}/item`)
  const responses = require(`./data/${diabloVersion}/responses`)
  const diabloData = {
    npcs: loadCsv(`${__dirname}/data/${diabloVersion}/monster_names.txt`, '\t', ['name']),
    areas: loadCsv(`${__dirname}/data/${diabloVersion}/areas.txt`, ',', ['name', 'id']),
    skills: loadCsv(`${__dirname}/data/${diabloVersion}/skills.txt`, '\t'),
    objects: loadCsv(`${__dirname}/data/${diabloVersion}/objects.txt`, '\t'),
    itemEnums: itemEnums,
    responses: responses,
    itemProperties: loadCsv(`${__dirname}/data/${diabloVersion}/item_properties.txt`, '|', ['statName', 'saveBits', 'saveParamBits', 'saveAdd']),
    itemData: loadCsv(`${__dirname}/data/${diabloVersion}/item_data.txt`, '|', ['name', 'code', 'classificationString', 'width', 'height', 'stackable', 'usable', 'throwable']),
    itemTypes: loadCsv(`${__dirname}/data/${diabloVersion}/itemtypes.txt`, '\t'),
    qualityItems: loadCsv(`${__dirname}/data/${diabloVersion}/qualityitems.txt`, '\t'),
    uniqueItems: loadCsv(`${__dirname}/data/${diabloVersion}/uniqueitems.txt`, '\t'),
    weapons: loadCsv(`${__dirname}/data/${diabloVersion}/weapons.txt`, '\t'),
    superUniques: loadCsv(`${__dirname}/data/${diabloVersion}/superuniques.txt`, '\t'),
    setItems: loadCsv(`${__dirname}/data/${diabloVersion}/setitems.txt`, '\t'),
    runes: loadCsv(`${__dirname}/data/${diabloVersion}/setitems.txt`, '\t'),
    misc: loadCsv(`${__dirname}/data/${diabloVersion}/misc.txt`, '\t'),
    monStats: loadCsv(`${__dirname}/data/${diabloVersion}/monstats.txt`, '\t'),
    missiles: loadCsv(`${__dirname}/data/${diabloVersion}/missiles.txt`, '\t'),
    armor: loadCsv(`${__dirname}/data/${diabloVersion}/armor.txt`, '\t'),
    gems: loadCsv(`${__dirname}/data/${diabloVersion}/gems.txt`, '\t'),
    belts: loadCsv(`${__dirname}/data/${diabloVersion}/belts.txt`, '\t')
  }

  return diabloDataToNode(diabloData)
}


lib/loader.js
module.exports = diabloDataToNode

function diabloDataToNode (diabloData) {
  const indexes = require('./indexes.js')(diabloData)
  return {
    npcs: diabloData.npcs,
    npcsByName: indexes.npcsByName,

    areas: indexes.areasById,
    areasArray: diabloData.areas,
    areasByName: indexes.areasByName,

    skills: indexes.skillsById,
    skillsByName: indexes.skillsByName,
    skillsArray: diabloData.npcs,

    objects: indexes.objectsById,
    objectsByName: indexes.objectsByName,
    objectsArray: diabloData.objects,

    itemEnums: diabloData.itemEnums,

    itemProperties: diabloData.itemProperties,
    itemPropertiesByName: indexes.itemPropertiesByName,

    itemData: diabloData.itemData,
    itemDataByName: indexes.itemDataByName,

    itemTypes: diabloData.itemTypes,
    itemTypesByName: indexes.itemTypesByName,

    qualityItems: diabloData.qualityItems,

    uniqueItems: diabloData.uniqueItems,
    uniqueItemsByName: indexes.uniqueItemsByName,

    weapons: diabloData.weapons,
    weaponsByName: indexes.weaponsByName,

    setItems: diabloData.setItems,
    setItemsByName: indexes.setItemsByName,

    runes: diabloData.runes,
    runesByName: indexes.runesByName,

    misc: diabloData.misc,
    miscByName: indexes.miscByName,

    monStats: diabloData.monStats,
    monStatsByName: indexes.monStatsByName,

    missiles: indexes.missilesById,
    missilesArray: diabloData.missiles,
    missilesByName: indexes.missilesByName,

    armor: diabloData.armor,
    armorByName: indexes.armorByName,

    gems: diabloData.gems,
    gemsByName: indexes.gemsByName,

    belts: diabloData.belts,
    beltsByName: indexes.beltsByName,

    responses: diabloData.responses
  }
}


lib/indexes.js
const indexer = require('./indexer.js')

module.exports = function (diabloData) {
  return {
    npcsByName: indexer.buildIndexFromArray(diabloData.npcs, 'name'),

    areasById: indexer.buildIndexFromArray(diabloData.areas, 'id'),
    areasByName: indexer.buildIndexFromArray(diabloData.areas, 'name'),

    skillsById: indexer.buildIndexFromArray(diabloData.skills, 'Id'),
    skillsByName: indexer.buildIndexFromArray(diabloData.skills, 'skill'),

    objectsById: indexer.buildIndexFromArray(diabloData.objects, 'Id'),
    objectsByName: indexer.buildIndexFromArray(diabloData.objects, 'Name'),

    itemPropertiesByName: indexer.buildIndexFromArray(diabloData.itemProperties, 'statName'),

    itemDataByName: indexer.buildIndexFromArray(diabloData.itemData, 'name'),

    itemTypesByName: indexer.buildIndexFromArray(diabloData.itemTypes, 'ItemType'),

    uniqueItemsByName: indexer.buildIndexFromArray(diabloData.uniqueItems, 'type'),

    weaponsByName: indexer.buildIndexFromArray(diabloData.weapons, 'name'),

    superUniquesByName: indexer.buildIndexFromArray(diabloData.superUniques, 'name'),

    setItemsByName: indexer.buildIndexFromArray(diabloData.setItems, 'name'),

    runesByName: indexer.buildIndexFromArray(diabloData.runes, 'Rune Name'),

    miscByName: indexer.buildIndexFromArray(diabloData.misc, 'name'),

    monStatsByName: indexer.buildIndexFromArray(diabloData.monStats, 'Class'),

    missilesById: indexer.buildIndexFromArray(diabloData.missiles, 'Id'),
    missilesByName: indexer.buildIndexFromArray(diabloData.missiles, 'Missile'),

    armorByName: indexer.buildIndexFromArray(diabloData.armor, 'name'),

    gemsByName: indexer.buildIndexFromArray(diabloData.gems, 'name'),

    beltsByName: indexer.buildIndexFromArray(diabloData.belts, 'name')
  }
}


data/pod_1.13d/responses.js
const joinGame = {
  0x00: 'Game joining succeeded. In this case, Diablo 2 terminates the connection with the MCP and initiates the connection with the D2GS.',
  0x29: 'Password incorrect.',
  0x2A: 'Game does not exist.',
  0x2B: 'Game is full.',
  0x2C: 'You do not meet the level requirements for this game.',
  0x6E: 'A dead hardcore character cannot join a game.',
  0x71: 'A non-hardcore character cannot join a game created by a Hardcore character.',
  0x73: 'Unable to join a Nightmare game.',
  0x74: 'Unable to join a Hell game.',
  0x78: 'A non-expansion character cannot join a game created by an Expansion character.',
  0x79: 'A Expansion character cannot join a game created by a non-expansion character.',
  0x7D: 'A non-ladder character cannot join a game created by a Ladder character.'
}

const probablyNoInternet = 'Realm Unavailable: No Battle.net connection detected.'
const startup = {
  0x00: 'Success',
  0x02: probablyNoInternet,
  0x0A: probablyNoInternet,
  0x0B: probablyNoInternet,
  0x0C: probablyNoInternet,
  0x0D: probablyNoInternet,
  0x7E: 'CDKey banned from realm play.',
  0x7F: 'Temporary IP ban "Your connection has been temporarily restricted from this realm. Please try to log in at another time"'
}

const logon = {
  0x00: 'Success',
  0x01: 'Account doesn\'t exist',
  0x02: 'Invalid password',
  0x03: 'Account closed'
}
module.exports = { joinGame, startup, logon }


data/pod_1.13d/item.js
function keysToObject (keys) {
  return keys.reduce((acc, e, i) => {
    acc[e] = i
    return acc
  }, {})
}

const VersionType =
{
  classic: 0,
  classic110: 2,
  lod: 100,
  lod110: 101
}

const ClassType = keysToObject(
  [
    'not_applicable',
    'normal',
    'exceptional',
    'elite'
  ])

const QualityType = keysToObject(
  [
    'not_applicable',
    'inferior',
    'normal',
    'superior',
    'magical',
    'set',
    'rare',
    'unique',
    'crafted'
  ])

const DirectoryType =
{
  not_applicable: 0,
  helm: 1,
  amulet: 2,
  armor: 3,
  right_hand: 4,
  left_hand: 5,
  right_hand_ring: 6,
  left_hand_ring: 7,
  belt: 8,
  boots: 9,
  gloves: 10,
  right_hand_switch: 11,
  left_hand_switch: 12
}

const ContainerType =
{
  unspecified: 0,
  inventory: 2, // 10x8 size
  inventory_bottom: 3, // bottom is size 10x1 (just last line ...)
  trader_offer: 4,
  for_trade: 6,
  cube: 8,
  stash: 0x0A, // 10x8
  stash_bottom: 0x0B, // 10x2
  belt: 0x20,
  item: 0x40,
  armor_tab: 0x82,
  weapon_tab_1: 0x84,
  weapon_tab_2: 0x86,
  misc_tab: 0x88
}

const Action =
{
  add_to_ground: 0,
  ground_to_cursor: 1,
  drop_to_ground: 2,
  on_ground: 3,
  put_in_container: 4,
  remove_from_container: 5,
  equip: 6,
  indirectly_swap_body_item: 7,
  unequip: 8,
  swap_body_item: 9,
  add_quantity: 0x0A,
  add_to_shop: 0x0B,
  remove_from_shop: 0x0C,
  swap_in_container: 0x0D,
  put_in_belt: 0x0E,
  remove_from_belt: 0x0F,
  swap_in_belt: 0x10,
  auto_unequip: 0x11,
  to_cursor: 0x12,
  item_in_socket: 0x13,
  update_stats: 0x15,
  weapon_switch: 0x17
}

const itemContainerGcType =
{
  inventory: 0,
  trade: 2,
  cube: 3,
  stash: 4
}

const itemDestinationType =
{
  unspecified: 0,
  equipment: 1,
  belt: 2,
  ground: 3,
  cursor: 4,
  item: 6
}

const SuperiorItemClassType =
{
  not_applicable: -1,
  attackRating: 0,
  damage: 1,
  defense: 2,
  damage_attack_rating: 3,
  durability: 4,
  attack_rating_durability: 5,
  damage_durability: 6,
  defense_durability: 7
}

const ClassificationType = keysToObject(
  [
    'amazon_bow',
    'amazon_javelin',
    'amazon_spear',
    'amulet',
    'antidote_potion',
    'armor',
    'arrows',
    'assassin_katar',
    'axe',
    'barbarian_helm',
    'belt',
    'body_part',
    'bolts',
    'boots',
    'bow',
    'circlet',
    'club',
    'crossbow',
    'dagger',
    'druid_pelt',
    'ear',
    'elixir',
    'gem',
    'gloves',
    'gold',
    'grand_charm',
    'hammer',
    'health_potion',
    'helm',
    'herb',
    'javelin',
    'jewel',
    'key',
    'large_charm',
    'mace',
    'mana_potion',
    'necromancer_shrunken_head',
    'paladin_shield',
    'polearm',
    'quest_item',
    'rejuvenation_potion',
    'ring',
    'rune',
    'scepter',
    'scroll',
    'shield',
    'small_charm',
    'sorceress_orb',
    'spear',
    'staff',
    'stamina_potion',
    'sword',
    'thawing_potion',
    'throwing_axe',
    'throwing_knife',
    'throwing_potion',
    'tome',
    'torch',
    'wand'
  ])

const itemStatType = keysToObject(
  [
    'strength',
    'energy',
    'dexterity',
    'vitality',
    'all_attributes',
    'new_skills',
    'life',
    'maximum_life',
    'mana',
    'maximum_mana',
    'stamina',
    'maximum_stamina',
    'level',
    'experience',
    'gold',
    'bank',
    'enhanced_defense',
    'enhanced_maximum_damage',
    'enhanced_minimum_damage',
    'attack_rating',
    'increased_blocking',
    'minimum_damage',
    'maximum_damage',
    'secondary_minimum_damage',
    'secondary_maximum_damage',
    'enhanced_damage',
    'mana_recovery',
    'mana_recovery_bonus',
    'stamina_recovery_bonus',
    'last_experience',
    'next_experience',
    'defense',
    'defense_vs_missiles',
    'defense_vs_melee',
    'damage_reduction',
    'magical_damage_reduction',
    'damage_reduction_percent',
    'magical_damage_reduction_percent',
    'maximum_magical_damage_reduction_percent',
    'fire_resistance',
    'maximum_fire_resistance',
    'lightning_resistance',
    'maximum_lightning_resistance',
    'cold_resistance',
    'maximum_cold_resistance',
    'poison_resistance',
    'maximum_poison_resistance',
    'damage_aura',
    'minimum_fire_damage',
    'maximum_fire_damage',
    'minimum_lightning_damage',
    'maximum_lightning_damage',
    'minimum_magical_damage',
    'maximum_magical_damage',
    'minimum_cold_damage',
    'maximum_cold_damage',
    'cold_damage_length',
    'minimum_poison_damage',
    'maximum_poison_damage',
    'poison_damage_length',
    'minimum_life_stolen_per_hit',
    'maximum_life_stolen_per_hit',
    'minimum_mana_stolen_per_hit',
    'maximum_mana_stolen_per_hit',
    'minimum_stamina_drain',
    'maximum_stamina_drain',
    'stun_length',
    'velocity_percent',
    'attack_rate',
    'other_animation_rate',
    'quantity',
    'value',
    'durability',
    'maximum_durability',
    'replenish_life',
    'enhanced_maximum_durability',
    'enhanced_life',
    'enhanced_mana',
    'attacker_takes_damage',
    'extra_gold',
    'better_chance_of_getting_magic_item',
    'knockback',
    'time_duration',
    'class_skills',
    'unsent_parameter',
    'add_experience',
    'life_after_each_kill',
    'reduce_vendor_prices',
    'double_herb_duration',
    'light_radius',
    'light_colour',
    'reduced_requirements',
    'reduced_level_requirement',
    'increased_attack_speed',
    'reduced_level_requirement_percent',
    'last_block_frame',
    'faster_run_walk',
    'non_class_skill',
    'state',
    'faster_hit_recovery',
    'monster_player_count',
    'skill_poison_override_length',
    'faster_block_rate',
    'skill_bypass_undead',
    'skill_bypass_demons',
    'faster_cast_rate',
    'skill_bypass_beasts',
    'single_skill',
    'slain_monsters_rest_in_peace',
    'curse_resistance',
    'poison_length_reduction',
    'adds_damage',
    'hit_causes_monster_to_flee',
    'hit_blinds_target',
    'damage_to_mana',
    'ignore_targets_defense',
    'reduce_targets_defense',
    'prevent_monster_heal',
    'half_freeze_duration',
    'to_hit_percent',
    'monster_defense_reduction_per_hit',
    'damage_to_demons',
    'damage_to_undead',
    'attack_rating_against_demons',
    'attack_rating_against_undead',
    'throwable',
    'elemental_skills',
    'all_skills',
    'attacker_takes_lightning_damage',
    'iron_maiden_level',
    'lifetap_level',
    'thorns_percent',
    'bone_armor',
    'maximum_bone_armor',
    'freezes_target',
    'open_wounds',
    'crushing_blow',
    'kick_damage',
    'mana_after_each_kill',
    'life_after_each_demon_kill',
    'extra_blood',
    'deadly_strike',
    'fire_absorption_percent',
    'fire_absorption',
    'lightning_absorption_percent',
    'lightning_absorption',
    'magic_absorption_percent',
    'magic_absorption',
    'cold_absorption_percent',
    'cold_absorption',
    'slows_down_enemies',
    'aura',
    'indestructible',
    'cannot_be_frozen',
    'stamina_drain_percent',
    'reanimate',
    'piercing_attack',
    'fires_magic_arrows',
    'fire_explosive_arrows',
    'minimum_throwing_damage',
    'maximum_throwing_damage',
    'skill_hand_of_athena',
    'skill_stamina_percent',
    'skill_passive_stamina_percent',
    'concentration',
    'enchant',
    'pierce',
    'conviction',
    'chilling_armor',
    'frenzy',
    'decrepify',
    'skill_armor_percent',
    'alignment',
    'target_0',
    'target_1',
    'gold_lost',
    'conversion_level',
    'conversion_maximum_life',
    'unit_do_overlay',
    'attack_rating_against_monster_type',
    'damage_to_monster_type',
    'fade',
    'armor_override_percent',
    'unused_183',
    'unused_184',
    'unused_185',
    'unused_186',
    'unused_187',
    'skill_tab',
    'unused_189',
    'unused_190',
    'unused_191',
    'unused_192',
    'unused_193',
    'socket_count',
    'skill_on_striking',
    'skill_on_kill',
    'skill_on_death',
    'skill_on_hit',
    'skill_on_level_up',
    'unused_200',
    'skill_when_struck',
    'unused_202',
    'unused_203',
    'charged',
    'unused_204',
    'unused_205',
    'unused_206',
    'unused_207',
    'unused_208',
    'unused_209',
    'unused_210',
    'unused_211',
    'unused_212',
    'defense_per_level',
    'enhanced_defense_per_level',
    'life_per_level',
    'mana_per_level',
    'maximum_damage_per_level',
    'maximum_enhanced_damage_per_level',
    'strength_per_level',
    'dexterity_per_level',
    'energy_per_level',
    'vitality_per_level',
    'attack_rating_per_level',
    'bonus_to_attack_rating_per_level',
    'maximum_cold_damage_per_level',
    'maximum_fire_damage_per_level',
    'maximum_lightning_damage_per_level',
    'maximum_poison_damage_per_level',
    'cold_resistance_per_level',
    'fire_resistance_per_level',
    'lightning_resistance_per_level',
    'poison_resistance_per_level',
    'cold_absorption_per_level',
    'fire_absorption_per_level',
    'lightning_absorption_per_level',
    'poison_absorption_per_level',
    'thorns_per_level',
    'extra_gold_per_level',
    'better_chance_of_getting_magic_item_per_level',
    'stamina_regeneration_per_level',
    'stamina_per_level',
    'damage_to_demons_per_level',
    'damage_to_undead_per_level',
    'attack_rating_against_demons_per_level',
    'attack_rating_against_undead_per_level',
    'crushing_blow_per_level',
    'open_wounds_per_level',
    'kick_damage_per_level',
    'deadly_strike_per_level',
    'find_gems_per_level',
    'repairs_durability',
    'replenishes_quantity',
    'increased_stack_size',
    'find_item',
    'slash_damage',
    'slash_damage_percent',
    'crush_damage',
    'crush_damage_percent',
    'thrust_damage',
    'thrust_damage_percent',
    'slash_damage_absorption',
    'crush_damage_absorption',
    'thrust_damage_absorption',
    'slash_damage_absorption_percent',
    'crush_damage_absorption_percent',
    'thrust_damage_absorption_percent',
    'defense_per_time',
    'enhanced_defense_per_time',
    'life_per_time',
    'mana_per_time',
    'maximum_damage_per_time',
    'maximum_enhanced_damage_per_time',
    'strength_per_time',
    'dexterity_per_time',
    'energy_per_time',
    'vitality_per_time',
    'attack_rating_per_time',
    'chance_to_hit_per_time',
    'maximum_cold_damage_per_time',
    'maximum_fire_damage_per_time',
    'maximum_lightning_damage_per_time',
    'maximum_damage_per_poison',
    'cold_resistance_per_time',
    'fire_resistance_per_time',
    'lightning_resistance_per_time',
    'poison_resistance_per_time',
    'cold_absorption_per_time',
    'fire_absorption_per_time',
    'lightning_absorption_per_time',
    'poison_absorption_per_time',
    'extra_gold_per_time',
    'better_chance_of_getting_magic_item_per_time',
    'regenerate_stamina_per_time',
    'stamina_per_time',
    'damage_to_demons_per_time',
    'damage_to_undead_per_time',
    'attack_rating_against_demons_per_time',
    'attack_rating_against_undead_per_time',
    'crushing_blow_per_time',
    'open_wounds_per_time',
    'kick_damage_per_time',
    'deadly_strike_per_time',
    'find_gems_per_time',
    'enemy_cold_resistance_reduction',
    'enemy_fire_resistance_reduction',
    'enemy_lightning_resistance_reduction',
    'enemy_poison_resistance_reduction',
    'damage_vs_monsters',
    'enhanced_damage_vs_monsters',
    'attack_rating_against_monsters',
    'bonus_to_attack_rating_against_monsters',
    'defense_vs_monsters',
    'enhanced_defense_vs_monsters',
    'fire_damage_length',
    'minimum_fire_damage_length',
    'maximum_fire_damage_length',
    'progressive_damage',
    'progressive_steal',
    'progressive_other',
    'progressive_fire',
    'progressive_cold',
    'progressive_lightning',
    'extra_charges',
    'progressive_attack_rating',
    'poison_count',
    'damage_framerate',
    'pierce_idx',
    'fire_mastery',
    'lightning_mastery',
    'cold_mastery',
    'poison_mastery',
    'passive_enemy_fire_resistance_reduction',
    'passive_enemy_lightning_resistance_reduction',
    'passive_enemy_cold_resistance_reduction',
    'passive_enemy_poison_resistance_reduction',
    'critical_strike',
    'dodge',
    'avoid',
    'evade',
    'warmth',
    'melee_attack_rating_mastery',
    'melee_damage_mastery',
    'melee_critical_hit_mastery',
    'thrown_weapon_attack_rating_mastery',
    'thrown_weapon_damage_mastery',
    'thrown_weapon_critical_hit_mastery',
    'weapon_block',
    'summon_resist',
    'modifier_list_skill',
    'modifier_list_level',
    'last_sent_life_percent',
    'source_unit_type',
    'source_unit_id',
    'short_parameter_1',
    'quest_item_difficulty',
    'passive_magical_damage_mastery',
    'passive_magical_resistance_reduction'
  ])

function reverseObject (o) {
  return Object.keys(o).reduce((acc, k) => {
    acc[o[k]] = k
    return acc
  }, {})
}

const itemStatTypeReversed = reverseObject(itemStatType)

function isArmor (classification) {
  switch (classification) {
    case ClassificationType.helm:
    case ClassificationType.armor:
    case ClassificationType.shield:
    case ClassificationType.gloves:
    case ClassificationType.boots:
    case ClassificationType.belt:
    case ClassificationType.druid_pelt:
    case ClassificationType.barbarian_helm:
    case ClassificationType.paladin_shield:
    case ClassificationType.necromancer_shrunken_head:
    case ClassificationType.circlet:
      return true
  }
  return false
}
function isWeapon (classification) {
  switch (classification) {
    case ClassificationType.amazon_bow:
    case ClassificationType.amazon_javelin:
    case ClassificationType.amazon_spear:
    case ClassificationType.assassin_katar:
    case ClassificationType.axe:
    case ClassificationType.bow:
    case ClassificationType.club:
    case ClassificationType.crossbow:
    case ClassificationType.dagger:
    case ClassificationType.hammer:
    case ClassificationType.javelin:
    case ClassificationType.mace:
    case ClassificationType.polearm:
    case ClassificationType.scepter:
    case ClassificationType.sorceress_orb:
    case ClassificationType.spear:
    case ClassificationType.sword:
    case ClassificationType.staff:
    case ClassificationType.throwing_axe:
    case ClassificationType.throwing_knife:
    case ClassificationType.throwing_potion:
    case ClassificationType.wand:
      return true
  }
  return false
}

module.exports = { VersionType, ClassType, QualityType, DirectoryType, ContainerType, Action, itemContainerGcType, itemDestinationType, itemStatType, SuperiorItemClassType, ClassificationType, isArmor, isWeapon, itemStatTypeReversed }
